File: velocity/src/main/java/ru/privatenull/pnauth/velocity/VelocityPlayerDisplay.java
package ru.privatenull.pnauth.velocity;

import com.velocitypowered.api.proxy.Player;
import com.velocitypowered.api.proxy.ProxyServer;
import net.kyori.adventure.bossbar.BossBar;
import net.kyori.adventure.title.Title;
import ru.privatenull.pnauth.display.*;
import ru.privatenull.pnauth.message.MessageFormat;
import ru.privatenull.pnauth.platform.PlayerResourceKey;
import java.time.Duration;
import java.util.UUID;
import java.util.concurrent.*;

final class VelocityPlayerDisplay implements PlayerDisplay, AutoCloseable {
    private final ProxyServer proxy; private final MessageFormat format;
    private final ConcurrentMap<PlayerResourceKey,Action> actions=new ConcurrentHashMap<>();private final ConcurrentMap<PlayerResourceKey,Titles> titles=new ConcurrentHashMap<>();private final ConcurrentMap<PlayerResourceKey,Boss> bosses=new ConcurrentHashMap<>();
    private final ScheduledExecutorService scheduler=Executors.newScheduledThreadPool(2,r->{Thread t=new Thread(r,"pnauth-velocity-display");t.setDaemon(true);return t;});
    VelocityPlayerDisplay(ProxyServer proxy,MessageFormat format){this.proxy=proxy;this.format=format;}
    public ActionBarHandle actionBar(UUID id,String name,ActionBarOptions o){PlayerResourceKey k=key(id,name);return actions.compute(k,(x,v)->{if(v==null||!v.active)return new Action(k,o);v.apply(o);return v;});}
    public TitleHandle title(UUID id,String name,TitleOptions o){PlayerResourceKey k=key(id,name);return titles.compute(k,(x,v)->{if(v==null||!v.active)return new Titles(k,o);v.apply(o);return v;});}
    public BossBarHandle bossBar(UUID id,String name,BossBarOptions o){PlayerResourceKey k=key(id,name);return bosses.compute(k,(x,v)->{if(v==null||!v.active)return new Boss(k,o);v.apply(o);return v;});}
    public java.util.Optional<ActionBarHandle> findActionBar(UUID id,String name){return java.util.Optional.ofNullable(actions.get(key(id,name)));}public java.util.Optional<TitleHandle> findTitle(UUID id,String name){return java.util.Optional.ofNullable(titles.get(key(id,name)));}public java.util.Optional<BossBarHandle> findBossBar(UUID id,String name){return java.util.Optional.ofNullable(bosses.get(key(id,name)));}
    public boolean removeActionBar(UUID id,String name){Action h=actions.remove(key(id,name));if(h!=null)h.close();return h!=null;}public boolean removeTitle(UUID id,String name){Titles h=titles.remove(key(id,name));if(h!=null)h.close();return h!=null;}public boolean removeBossBar(UUID id,String name){Boss h=bosses.remove(key(id,name));if(h!=null)h.close();return h!=null;}
    public void clear(UUID id){actions.entrySet().stream().filter(e->e.getKey().playerId.equals(id)).map(java.util.Map.Entry::getValue).toList().forEach(Action::close);titles.entrySet().stream().filter(e->e.getKey().playerId.equals(id)).map(java.util.Map.Entry::getValue).toList().forEach(Titles::close);bosses.entrySet().stream().filter(e->e.getKey().playerId.equals(id)).map(java.util.Map.Entry::getValue).toList().forEach(Boss::close);}
    public void close(){actions.values().forEach(Action::close);titles.values().forEach(Titles::close);bosses.values().forEach(Boss::close);scheduler.shutdownNow();}
    private abstract class Base implements DisplayHandle{
        final PlayerResourceKey key;final UUID playerId;volatile boolean active=true,paused;volatile ScheduledFuture<?> expiry;
        Base(PlayerResourceKey key,Duration lifetime){this.key=key;playerId=key.playerId;lifetime(lifetime);}public UUID playerId(){return playerId;}public String displayId(){return key.name;}public boolean active(){return active;}public boolean paused(){return paused;}
        public void pause(){paused=true;}public void resume(){if(active)paused=false;}public synchronized void lifetime(Duration value){cancel(expiry);Duration d=valid(value);if(!d.isZero())expiry=scheduler.schedule(this::close,d.toMillis(),TimeUnit.MILLISECONDS);}
        Player player(){return proxy.getPlayer(playerId).orElse(null);}void cancel(ScheduledFuture<?> task){if(task!=null)task.cancel(false);}
    }
    private final class Action extends Base implements ActionBarHandle{
        volatile String text;volatile Duration interval;volatile ScheduledFuture<?> refresh;
        Action(PlayerResourceKey key,ActionBarOptions o){super(key,o.lifetime());text=o.text();interval=o.refreshInterval();sendNow();reschedule();}void apply(ActionBarOptions o){text(o.text());refreshInterval(o.refreshInterval());lifetime(o.lifetime());}
        public void text(String value){text=value==null?"":value;sendNow();}public synchronized void refreshInterval(Duration value){interval=valid(value);reschedule();}
        private synchronized void reschedule(){cancel(refresh);if(active&&!interval.isZero())refresh=scheduler.scheduleAtFixedRate(this::sendNow,interval.toMillis(),interval.toMillis(),TimeUnit.MILLISECONDS);}
        public void sendNow(){Player p=player();if(active&&!paused&&p!=null)p.sendActionBar(VelocityMessages.component(text,format));}
        public synchronized void close(){if(!active)return;active=false;actions.remove(key,this);cancel(refresh);cancel(expiry);Player p=player();if(p!=null)p.sendActionBar(net.kyori.adventure.text.Component.empty());}
    }
    private final class Titles extends Base implements TitleHandle{
        volatile String title,subtitle;volatile Duration fadeIn,stay,fadeOut,repeat;volatile ScheduledFuture<?> refresh;
        Titles(PlayerResourceKey key,TitleOptions o){super(key,o.lifetime());title=o.title();subtitle=o.subtitle();fadeIn=o.fadeIn();stay=o.stay();fadeOut=o.fadeOut();repeat=o.repeatInterval();showNow();reschedule();}void apply(TitleOptions o){title=o.title();subtitle=o.subtitle();timings(o.fadeIn(),o.stay(),o.fadeOut());repeatInterval(o.repeatInterval());lifetime(o.lifetime());showNow();}
        public void title(String value){title=value==null?"":value;showNow();}public void subtitle(String value){subtitle=value==null?"":value;showNow();}public void timings(Duration a,Duration b,Duration c){fadeIn=valid(a);stay=valid(b);fadeOut=valid(c);}
        public synchronized void repeatInterval(Duration value){repeat=valid(value);reschedule();}private synchronized void reschedule(){cancel(refresh);if(active&&!repeat.isZero())refresh=scheduler.scheduleAtFixedRate(this::showNow,repeat.toMillis(),repeat.toMillis(),TimeUnit.MILLISECONDS);}
        public void showNow(){Player p=player();if(active&&!paused&&p!=null)p.showTitle(Title.title(VelocityMessages.component(title,format),VelocityMessages.component(subtitle,format),Title.Times.times(fadeIn,stay,fadeOut)));}
        public void clear(){Player p=player();if(p!=null)p.clearTitle();}public synchronized void close(){if(!active)return;active=false;titles.remove(key,this);cancel(refresh);cancel(expiry);clear();}
    }
    private final class Boss extends Base implements BossBarHandle{
        final BossBar bar;volatile float progress;volatile ScheduledFuture<?> animation;
        Boss(PlayerResourceKey key,BossBarOptions o){super(key,o.lifetime());progress=o.progress();bar=BossBar.bossBar(VelocityMessages.component(o.text(),format),progress,colorOf(o.color()),overlayOf(o.overlay()));properties(o.darkenScreen(),o.playBossMusic(),o.createWorldFog());Player p=player();if(p!=null)p.showBossBar(bar);}void apply(BossBarOptions o){text(o.text());progress(o.progress());color(o.color());overlay(o.overlay());properties(o.darkenScreen(),o.playBossMusic(),o.createWorldFog());lifetime(o.lifetime());}
        public void text(String value){bar.name(VelocityMessages.component(value,format));}public void progress(float value){progress=clamp(value);bar.progress(progress);}public void color(BossBarColor value){bar.color(colorOf(value));}public void overlay(BossBarOverlay value){bar.overlay(overlayOf(value));}
        public void properties(boolean dark,boolean music,boolean fog){java.util.Set<BossBar.Flag> flags=new java.util.HashSet<>();if(dark)flags.add(BossBar.Flag.DARKEN_SCREEN);if(music)flags.add(BossBar.Flag.PLAY_BOSS_MUSIC);if(fog)flags.add(BossBar.Flag.CREATE_WORLD_FOG);bar.flags(flags);}
        public synchronized void animateProgress(float target,Duration duration,Easing easing){cancel(animation);float start=progress,end=clamp(target);int steps=Math.max(1,(int)(valid(duration).toMillis()/50));java.util.concurrent.atomic.AtomicInteger step=new java.util.concurrent.atomic.AtomicInteger();Easing curve=easing==null?Easing.LINEAR:easing;animation=scheduler.scheduleAtFixedRate(()->{if(paused)return;float ratio=Math.min(1F,step.incrementAndGet()/(float)steps);progress(start+(end-start)*curve.apply(ratio));if(ratio>=1F)cancel(animation);},0,50,TimeUnit.MILLISECONDS);}
        @Override public void pause(){if(!active||paused)return;paused=true;Player p=player();if(p!=null)p.hideBossBar(bar);}@Override public void resume(){if(!active||!paused)return;paused=false;Player p=player();if(p!=null)p.showBossBar(bar);}
        public synchronized void close(){if(!active)return;active=false;bosses.remove(key,this);cancel(animation);cancel(expiry);Player p=player();if(p!=null)p.hideBossBar(bar);}
    }
    private static Duration valid(Duration value){return value==null||value.isNegative()?Duration.ZERO:value;}private static float clamp(float value){return Math.max(0F,Math.min(1F,value));}
    private static BossBar.Color colorOf(BossBarColor c){return BossBar.Color.valueOf((c==null?BossBarColor.PURPLE:c).name());}
    private static BossBar.Overlay overlayOf(BossBarOverlay o){return BossBar.Overlay.valueOf((o==null?BossBarOverlay.PROGRESS:o).name());}
    private static PlayerResourceKey key(UUID id,String name){return new PlayerResourceKey(id,name);}
}

File: velocity/src/main/kotlin/ru/privatenull/pnauth/velocity/VelocityPlayerDisplay.kt
package ru.privatenull.pnauth.velocity

import com.velocitypowered.api.proxy.Player
import com.velocitypowered.api.proxy.ProxyServer
import net.kyori.adventure.bossbar.BossBar
import net.kyori.adventure.text.Component
import net.kyori.adventure.title.Title
import ru.privatenull.pnauth.display.ActionBarHandle
import ru.privatenull.pnauth.display.ActionBarOptions
import ru.privatenull.pnauth.display.BossBarColor
import ru.privatenull.pnauth.display.BossBarHandle
import ru.privatenull.pnauth.display.BossBarOptions
import ru.privatenull.pnauth.display.BossBarOverlay
import ru.privatenull.pnauth.display.DisplayHandle
import ru.privatenull.pnauth.display.Easing
import ru.privatenull.pnauth.display.PlayerDisplay
import ru.privatenull.pnauth.display.TitleHandle
import ru.privatenull.pnauth.display.TitleOptions
import ru.privatenull.pnauth.message.MessageFormat
import ru.privatenull.pnauth.platform.PlayerResourceKey
import java.time.Duration
import java.util.Objects
import java.util.Optional
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

class VelocityPlayerDisplay(
    private val proxy: ProxyServer,
    private val format: MessageFormat
) : ru.privatenull.pnauth.platform.adapter.PlatformDisplayAdapter, AutoCloseable {

    private val titleAdapter = ru.privatenull.pnauth.velocity.display.VelocityTitleAdapter(proxy, format)

    override fun showTitle(uniqueId: UUID, builder: ru.privatenull.pnauth.display.TitleBuilder) {
        titleAdapter.showTitle(uniqueId, builder)
    }

    override fun clearTitle(uniqueId: UUID) {
        titleAdapter.clearTitle(uniqueId)
    }

    private val actions = ConcurrentHashMap<PlayerResourceKey, Action>()
    private val titles = ConcurrentHashMap<PlayerResourceKey, Titles>()
    private val bosses = ConcurrentHashMap<PlayerResourceKey, Boss>()
    private val scheduler: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor { runnable ->
        val thread = Thread(runnable, "pnauth-display")
        thread.isDaemon = true
        thread
    }

    override fun actionBar(playerId: UUID, displayId: String, options: ActionBarOptions): ActionBarHandle {
        val key = PlayerResourceKey(playerId, displayId)
        val handle = actions.computeIfAbsent(key) { Action(key, options) }
        handle.applyOptions(options)
        return handle
    }

    override fun title(playerId: UUID, displayId: String, options: TitleOptions): TitleHandle {
        val key = PlayerResourceKey(playerId, displayId)
        val handle = titles.computeIfAbsent(key) { Titles(key, options) }
        handle.applyOptions(options)
        return handle
    }

    override fun bossBar(playerId: UUID, displayId: String, options: BossBarOptions): BossBarHandle {
        val key = PlayerResourceKey(playerId, displayId)
        val handle = bosses.computeIfAbsent(key) { Boss(key, options) }
        handle.applyOptions(options)
        return handle
    }

    override fun findActionBar(playerId: UUID, displayId: String): Optional<ActionBarHandle> {
        return Optional.ofNullable(actions[PlayerResourceKey(playerId, displayId)])
    }

    override fun findTitle(playerId: UUID, displayId: String): Optional<TitleHandle> {
        return Optional.ofNullable(titles[PlayerResourceKey(playerId, displayId)])
    }

    override fun findBossBar(playerId: UUID, displayId: String): Optional<BossBarHandle> {
        return Optional.ofNullable(bosses[PlayerResourceKey(playerId, displayId)])
    }

    override fun removeActionBar(playerId: UUID, displayId: String): Boolean {
        val removed = actions.remove(PlayerResourceKey(playerId, displayId))
        if (removed != null) {
            removed.close()
            return true
        }
        return false
    }

    override fun removeTitle(playerId: UUID, displayId: String): Boolean {
        val removed = titles.remove(PlayerResourceKey(playerId, displayId))
        if (removed != null) {
            removed.close()
            return true
        }
        return false
    }

    override fun removeBossBar(playerId: UUID, displayId: String): Boolean {
        val removed = bosses.remove(PlayerResourceKey(playerId, displayId))
        if (removed != null) {
            removed.close()
            return true
        }
        return false
    }

    override fun clear(playerId: UUID) {
        actions.keys.filter { it.playerId == playerId }.forEach { removeActionBar(it.playerId, it.name) }
        titles.keys.filter { it.playerId == playerId }.forEach { removeTitle(it.playerId, it.name) }
        bosses.keys.filter { it.playerId == playerId }.forEach { removeBossBar(it.playerId, it.name) }
    }

    override fun close() {
        actions.values.forEach(Action::close)
        titles.values.forEach(Titles::close)
        bosses.values.forEach(Boss::close)
        scheduler.shutdownNow()
    }

    private abstract inner class Base(
        val key: PlayerResourceKey,
        lifetime: Duration?
    ) : DisplayHandle {
        val playerId: UUID = key.playerId
        @Volatile protected var isCompletedActive: Boolean = true
        @Volatile protected var isCompletedPaused: Boolean = false
        @Volatile var expiry: ScheduledFuture<*>? = null

        init {
            lifetime(lifetime ?: Duration.ZERO)
        }

        override fun playerId(): UUID = playerId
        override fun displayId(): String = key.name
        override fun active(): Boolean = isCompletedActive
        override fun paused(): Boolean = isCompletedPaused

        override fun pause() {
            isCompletedPaused = true
        }

        override fun resume() {
            if (isCompletedActive) isCompletedPaused = false
        }

        @Synchronized
        override fun lifetime(lifetime: Duration) {
            cancel(expiry)
            val d = valid(lifetime)
            if (!d.isZero) {
                expiry = scheduler.schedule({ close() }, d.toMillis(), TimeUnit.MILLISECONDS)
            }
        }

        fun player(): Player? = proxy.getPlayer(playerId).orElse(null)

        fun cancel(task: ScheduledFuture<*>?) {
            task?.cancel(false)
        }
    }

    private inner class Action(
        key: PlayerResourceKey,
        o: ActionBarOptions
    ) : Base(key, o.lifetime()), ActionBarHandle {

        @Volatile private var valueText: String = o.text()
        @Volatile private var interval: Duration = valid(o.refreshInterval())
        @Volatile private var refresh: ScheduledFuture<*>? = null

        init {
            sendNow()
            reschedule()
        }

        fun applyOptions(o: ActionBarOptions) {
            text(o.text())
            refreshInterval(o.refreshInterval())
            lifetime(o.lifetime())
        }

        override fun text(text: String) {
            valueText = text
            sendNow()
        }

        @Synchronized
        override fun refreshInterval(interval: Duration) {
            this.interval = valid(interval)
            reschedule()
        }

        @Synchronized
        private fun reschedule() {
            cancel(refresh)
            if (isCompletedActive && !interval.isZero) {
                refresh = scheduler.scheduleAtFixedRate(
                    { sendNow() }, interval.toMillis(), interval.toMillis(), TimeUnit.MILLISECONDS
                )
            }
        }

        override fun sendNow() {
            val p = player()
            if (isCompletedActive && !isCompletedPaused && p != null) {
                p.sendActionBar(VelocityMessages.component(valueText, format))
            }
        }

        @Synchronized
        override fun close() {
            if (!isCompletedActive) return
            isCompletedActive = false
            actions.remove(key, this)
            cancel(refresh)
            cancel(expiry)
            val p = player()
            p?.sendActionBar(Component.empty())
        }
    }

    private inner class Titles(
        key: PlayerResourceKey,
        o: TitleOptions
    ) : Base(key, o.lifetime()), TitleHandle {

        @Volatile private var valueTitle: String = o.title()
        @Volatile private var valueSubtitle: String = o.subtitle()
        @Volatile private var fadeIn: Duration = valid(o.fadeIn())
        @Volatile private var stay: Duration = valid(o.stay())
        @Volatile private var fadeOut: Duration = valid(o.fadeOut())
        @Volatile private var repeat: Duration = valid(o.repeatInterval())
        @Volatile private var refresh: ScheduledFuture<*>? = null

        init {
            showNow()
            reschedule()
        }

        fun applyOptions(o: TitleOptions) {
            valueTitle = o.title()
            valueSubtitle = o.subtitle()
            timings(o.fadeIn(), o.stay(), o.fadeOut())
            repeatInterval(o.repeatInterval())
            lifetime(o.lifetime())
            showNow()
        }

        @Synchronized
        override fun title(title: String) {
            valueTitle = title
            showNow()
        }

        @Synchronized
        override fun subtitle(subtitle: String) {
            valueSubtitle = subtitle
            showNow()
        }

        @Synchronized
        override fun timings(fadeIn: Duration, stay: Duration, fadeOut: Duration) {
            this.fadeIn = valid(fadeIn)
            this.stay = valid(stay)
            this.fadeOut = valid(fadeOut)
        }

        @Synchronized
        override fun repeatInterval(interval: Duration) {
            repeat = valid(interval)
            reschedule()
        }

        @Synchronized
        private fun reschedule() {
            cancel(refresh)
            if (isCompletedActive && !repeat.isZero) {
                refresh = scheduler.scheduleAtFixedRate(
                    { showNow() }, repeat.toMillis(), repeat.toMillis(), TimeUnit.MILLISECONDS
                )
            }
        }

        @Synchronized
        override fun showNow() {
            val p = player()
            if (isCompletedActive && !isCompletedPaused && p != null) {
                p.showTitle(
                    Title.title(
                        VelocityMessages.component(valueTitle, format),
                        VelocityMessages.component(valueSubtitle, format),
                        Title.Times.times(fadeIn, stay, fadeOut)
                    )
                )
            }
        }

        override fun clear() {
            player()?.clearTitle()
        }

        @Synchronized
        override fun release() {
            if (!isCompletedActive) return
            isCompletedActive = false
            titles.remove(key, this)
            cancel(refresh)
            cancel(expiry)
        }

        @Synchronized
        override fun close() {
            if (!isCompletedActive) return
            release()
            clear()
        }
    }

    private inner class Boss(
        key: PlayerResourceKey,
        o: BossBarOptions
    ) : Base(key, o.lifetime()), BossBarHandle {

        private val bar: BossBar = BossBar.bossBar(
            VelocityMessages.component(o.text(), format),
            clamp(o.progress()),
            colorOf(o.color()),
            overlayOf(o.overlay())
        )
        @Volatile private var valueText: String = o.text()
        @Volatile private var progress: Float = clamp(o.progress())
        @Volatile private var animation: ScheduledFuture<*>? = null
        @Volatile private var subscribedPlayer: Player? = null

        init {
            properties(o.darkenScreen(), o.playBossMusic(), o.createWorldFog())
            showNow()
        }

        fun applyOptions(o: BossBarOptions) {
            text(o.text())
            progress(o.progress())
            color(o.color())
            overlay(o.overlay())
            properties(o.darkenScreen(), o.playBossMusic(), o.createWorldFog())
            lifetime(o.lifetime())
        }

        override fun text(text: String) {
            valueText = text
            bar.name(VelocityMessages.component(valueText, format))
        }

        override fun progress(progress: Float) {
            this.progress = clamp(progress)
            bar.progress(this.progress)
        }

        override fun color(color: BossBarColor) {
            bar.color(colorOf(color))
        }

        override fun overlay(overlay: BossBarOverlay) {
            bar.overlay(overlayOf(overlay))
        }

        override fun properties(darkenScreen: Boolean, playBossMusic: Boolean, createWorldFog: Boolean) {
            val flags = mutableSetOf<BossBar.Flag>()
            if (darkenScreen) flags.add(BossBar.Flag.DARKEN_SCREEN)
            if (playBossMusic) flags.add(BossBar.Flag.PLAY_BOSS_MUSIC)
            if (createWorldFog) flags.add(BossBar.Flag.CREATE_WORLD_FOG)
            bar.flags(flags)
        }

        @Synchronized
        override fun animateProgress(target: Float, duration: Duration, easing: Easing) {
            cancel(animation)
            val start = progress
            val end = clamp(target)
            val steps = maxOf(1, (valid(duration).toMillis() / 50).toInt())
            val step = AtomicInteger()
            val curve = easing
            animation = scheduler.scheduleAtFixedRate({
                if (paused()) return@scheduleAtFixedRate
                val ratio = minOf(1f, step.incrementAndGet() / steps.toFloat())
                progress(start + (end - start) * curve.apply(ratio))
                if (ratio >= 1f) cancel(animation)
            }, 0, 50, TimeUnit.MILLISECONDS)
        }

        override fun pause() {
            if (!isCompletedActive || isCompletedPaused) return
            super.pause()
            subscribedPlayer?.hideBossBar(bar)
        }

        override fun resume() {
            if (!isCompletedActive || !isCompletedPaused) return
            super.resume()
            showNow()
        }

        @Synchronized
        private fun showNow() {
            val p = player()
            if (isCompletedActive && !isCompletedPaused && p != null) {
                if (subscribedPlayer != p) {
                    subscribedPlayer?.hideBossBar(bar)
                    subscribedPlayer = p
                }
                p.showBossBar(bar)
            }
        }

        @Synchronized
        override fun close() {
            if (!isCompletedActive) return
            isCompletedActive = false
            bosses.remove(key, this)
            cancel(animation)
            cancel(expiry)
            subscribedPlayer?.hideBossBar(bar)
            subscribedPlayer = null
        }
    }

    companion object {
        private fun valid(value: Duration?): Duration = if (value == null || value.isNegative) Duration.ZERO else value

        private fun clamp(value: Float): Float = value.coerceIn(0f, 1f)

        private fun colorOf(color: BossBarColor?): BossBar.Color {
            return when (color ?: BossBarColor.PURPLE) {
                BossBarColor.PINK -> BossBar.Color.PINK
                BossBarColor.BLUE -> BossBar.Color.BLUE
                BossBarColor.RED -> BossBar.Color.RED
                BossBarColor.GREEN -> BossBar.Color.GREEN
                BossBarColor.YELLOW -> BossBar.Color.YELLOW
                BossBarColor.PURPLE -> BossBar.Color.PURPLE
                BossBarColor.WHITE -> BossBar.Color.WHITE
            }
        }

        private fun overlayOf(overlay: BossBarOverlay?): BossBar.Overlay {
            return when (overlay ?: BossBarOverlay.PROGRESS) {
                BossBarOverlay.PROGRESS -> BossBar.Overlay.PROGRESS
                BossBarOverlay.NOTCHED_6 -> BossBar.Overlay.NOTCHED_6
                BossBarOverlay.NOTCHED_10 -> BossBar.Overlay.NOTCHED_10
                BossBarOverlay.NOTCHED_12 -> BossBar.Overlay.NOTCHED_12
                BossBarOverlay.NOTCHED_20 -> BossBar.Overlay.NOTCHED_20
            }
        }
    }
}
